Allow dropping an image on an empty paragraph block to create an image block

kevin940726 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Allow dropping files/blocks/HTML on an empty paragraph block.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Fix https://github.com/WordPress/gutenberg/issues/29145.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

I also created the `dragFiles` util to help test DnD in Playwright e2e tests.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
1. Create a new post or edit an existing post.
2. Add an empty paragraph block (note that the default block in a newly created post is not a paragraph block, yet).
3. Drag an image to the block.
4. It should show the placeholder upon hovering.
5. Dropping the image should create an image block.

## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/7753001/191423167-52a86f37-a6ea-4c3f-9b71-0fdfdbf5f72c.mp4

kevin940726 committed:
Allow dropping an image on an empty paragraph block

kevin940726 commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -145,10 +123,58 @@ function ParagraphBlock( {
-			/>
+			>
+				{ ! content && (
+					<DropZone
```
The `<DropZone>` component is put inside the `<p>` tag to position itself. I'm not sure if this approach will impact the block's HTML tree though. An alternative is to use `<Popover>` to render the `<DropZone>` component instead.

talldan replied:
Hmm, yeah, not sure about the HTML changes here. Quite a lot of effort was put into making the paragraph block just a `<p>`

Could an option be to use `__experimentalUseDropZone` to avoid those changes?

I guess that means reimplementing quite a lot of the visual aspects of drop zone, unless that can be broken down into smaller parts. 🤔

kevin940726 replied:
Yeah, I just found out that changing the structure of the block breaks a bunch of tests. 😞

I'll see what I can do 🤔

kevin940726 commented on packages/components/src/drop-zone/index.js:
```diff
@@ -118,23 +118,26 @@ export default function DropZoneComponent( {
 	if ( isDraggingOverElement ) {
 		children = (
-			<motion.div
+			<motion.span
```
Since `<div>` cannot be put inside a `<p>` tag, change all the divs here to spans instead.

kevin940726 commented on packages/components/src/drop-zone/index.js:
```diff
@@ -118,23 +118,26 @@ export default function DropZoneComponent( {
-				<motion.div variants={ foreground }>
+				<motion.span
+					variants={ foreground }
+					className="components-drop-zone__content-foreground"
```
The foreground is here to automatically _squeeze_ the content into a single line when the height is not big enough.

kevin940726 commented on packages/components/src/drop-zone/style.scss:
```diff
@@ -8,10 +8,12 @@
 	&.is-active {
 		opacity: 1;
 		visibility: visible;
+		pointer-events: auto;
```
Only set `pointer-events` to `auto` when active(dragging) so that users can interact with the element beneath it.

kevin940726 commented on packages/e2e-test-utils-playwright/src/page-utils/drag-files.ts:
```diff
@@ -0,0 +1,131 @@
+		 * @param  x The X coordinate.
+		 * @param  y The Y coordinate.
+		 */
+		dragTo: async ( x: number, y: number ) => {
```
Ideally, we should allow `page.mouse.move()` and `page.mouse.up()` to map to these methods automatically. But unfortunately, it's not currently possible in Playwright.

ellatrix replied:
These work fine Puppeteer, can move the tests to Puppeteer for now?

kevin940726 replied:
I managed to make it work in Playwright too so I don't think it's necessary to move it to Puppeteer for now :).

talldan commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -108,6 +123,36 @@ function ParagraphBlock( {
 				</InspectorControls>
 			) }
+			{ ! content && (
+				<Popover
```
It might be good to extract this into a separate component.

talldan commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -108,6 +123,36 @@ function ParagraphBlock( {
+					__unstableForcePosition
+					className="wp-block-paragraph__drop_zone"
+				>
+					<DropZone
```
I do think it'd be interesting to try `useDropZone`, and only render the visual part of the dropzone when the user is actually dragging a file over.

That way you don't need the `pointer-events` change, and possibly also won't need to listen for resize of the paragraph.

talldan commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -108,6 +123,36 @@ function ParagraphBlock( {
+							height: paragraphElement?.offsetHeight,
+						} }
+						onFilesDrop={ ( files ) => {
+							if ( files.length === 1 ) {
```
I think there needs to be a file type check. Though it might be better to check for relevant `file` transforms and trigger the valid ones, so then gallery, video and file blocks will also be supported (as well as any third party blocks that register a valid transform).

talldan reviewed (commented):
> FWIW, I don't think this is the right path, but it fixes the issue, and it's always better than nothing 😅.
> 
> Ideally, we should be able to reuse useBlockDropZone or useOnBlockDrop in the block library, either that or handle all the dropping logic in the block editor package.

It'd be good to hear a proposal for that.

I wonder if there should be a basic transform system for this, similar to how the `files` transform already works. Maybe specifying a `from` and a `to` for a file transform allows dragging over a block and not just before and after it.

I don't think it's bad to start with something basic though and refactor it later. It gets the feature out there for testing before maturing it into a better API.

----

On another note, I noticed a small issue at the moment. It's questionable if it's worth solving. When you start a new post there isn't actually a paragraph block in the post.

There's 'Type /  to choose a block', but this is the default block appender masquerading as a paragraph, the paragraph is only created when that appender receives focus. I think it's done like this to avoid adding empty paragraphs into new posts (if the user decides to save without adding any content), but the problem is that the default block appender doesn't handle file drops the same way a paragraph does.

ciampo commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -108,6 +123,36 @@ function ParagraphBlock( {
 			) }
+			{ ! content && (
+				<Popover
+					anchorRef={ paragraphElement }
+					animate={ false }
+					position="top right left"
+					focusOnMount={ false }
+					__unstableForcePosition
```
Just a note, as we're currently reworking a few things around the `Popover` component:

- use the `placement` prop instead of `position`
- use the `anchor` prop instead of `anchorRef` (available after #43691 gets merged, hopefully next week)
- use the `flip` and `resize` props instead of `__unstableForcePosition`

mtias commented:
Thanks, this is a much better experience! I think it might be nice to not trigger the sibling inserter before an empty paragraph or empty media placeholder, since the effect is the same and it just adds a bit of cognitive overhead and precision issues.

@jasmussen @ellatrix would you mind checking this one when you have a moment?

jasmussen commented:
This is a really nice PR, and really improves the experience from what's shipping, where it feels weird that you can drop an image above and below a paragraph, but not _on_ a paragraph:

![before](https://user-images.githubusercontent.com/1204802/188405644-276b77fa-a577-418e-babd-da01361f7047.gif)

I'm seeing some positioning issues, however, running TT3. The dropzone appears to hover above the empty paragraph, and moves still further up the longer down the page it is:

![after](https://user-images.githubusercontent.com/1204802/188405877-3bec939c-3541-4f40-b726-695146ca55e6.gif)

I would echo Matías that if we can get this one working well, there isn't a need for a sibling inserter before and after an empty paragraph. As also briefly shown in the GIF above, the fact that those are present also, seems to interfere with the positioning of elements, so the code could potentially be simpler.

Finally on the visuals, and these aren't strong opinions so in principle we could land this as is, but nevertheless some thoughts:

* We con't use any upload text for the sibling inserter, whether dropping one or multiple images. So it doesn't feel like we need any here either — in that light, it could just be the a blue box that covers the entire paragraph. We could show an icon if we need to, but it seems better to not show anything, than show cropped text/icons.
* If need be, we can show text and/or icons if there's space. Global Styles would let you set both a big and a small font size, and in some cases a paragraph might be of varying sizes.  That's worth considering for the implementation as well.

A small note, it looks like you're testing with TT1. While the code should work with all themes, it's worth noting that TT1 comes with a great deal of custom editor CSS that tweaks margins and paddings, so it's important to also test in something like TT3 or Empty Theme, to make sure the code doesn't get tailored to one theme or the other.

I'll spin up a separate PR adding a 2px border radius to the dropzone, I just noticed we're missing that.

kevin940726 commented:
Thanks for trying it out! I'm currently busy helping out with other things. I'll get back to it soon.

kevin940726 committed:
Update popover props for best practice

kevin940726 committed:
Refactor

kevin940726 commented:
I refactored it to use `useOnBlockDrop` to implement the drop-zone. Haven't yet updated the tests and cleaned up the mess but LMK what you think!

https://user-images.githubusercontent.com/7753001/190333398-6b42e10d-4365-43df-9439-147010d4d122.mp4

I'm using the empty theme in the video above, but the paragraph block is a bit too small. The drop zone seems weird, I wonder if we should enlarge it a little bit?

kevin940726 committed:
Take reduced motion into account

jasmussen commented:
I'm seeing perfect positioning, the right border radius and color, and the right width/height, and some nice animation:

![status](https://user-images.githubusercontent.com/1204802/190362048-2004c962-e916-4aca-a9b2-2b8d2312573e.gif)

Very cool, this is close! Right now it looks like it's scaling both x and y axes. Can we scale only the y axis, so it starts as a full-width line that grows thicker?

That might thread the needle! Looks good!

mtias commented:
This is looking pretty good!

kevin940726 committed:
Update animation to scaleY only

talldan commented on packages/block-editor/src/components/use-block-drop-zone/index.js:
```diff
@@ -130,8 +130,16 @@ export default function useBlockDropZone( {
 
-			if ( targetIndex !== null ) {
-				showInsertionPoint( targetRootClientId, targetIndex );
+			const nearestBlock = blockElements[ targetIndex ];
+			if ( nearestBlock ) {
+				const previousBlock = blockElements[ targetIndex - 1 ];
+
+				if (
+					nearestBlock.dataset.empty !== 'true' &&
+					previousBlock?.dataset.empty !== 'true'
+				) {
+					showInsertionPoint( targetRootClientId, targetIndex );
+				}
```
Would be good to double-check what the expected behavior here is when dragging blocks normally, as at the moment this prevents the sibling inserter.

My instinct is it should only prevent the sibling inserter when dragging a file, but it's definitely tricky to build this in a generic way. 

Potentially we need a way for a block to declare that it's taking over as a drop target based on what's being dragged. It might be useful for dragging onto an image block placeholder as well. Worth keeping this simple for now though and following up.

kevin940726 replied:
Good point! It's a side effect of reusing the `useOnBlockDrop` hook here, but it might not be a bad one.

My understanding is that the empty paragraph block itself can be viewed as a dropping indicator since it has little semantic value. But it might not be ideal for some other cases.

This doesn't work great now when dropping a block onto a paragraph block though, the block gets inserted rather than replacing the empty paragraph block. I'll try to fix it or disable it altogether.

talldan commented on packages/block-editor/src/components/use-on-block-drop/index.js:
```diff
@@ -166,88 +109,134 @@ export function onFilesDrop(
-				transform.type === 'files' &&
-				canInsertBlockType( transform.blockName, targetRootClientId ) &&
-				transform.isMatch( files )
+				transform.type === 'files' && transform.isMatch( files )
```
What's the reasoning behind removing `canInsertBlockType` here?

kevin940726 replied:
Oops, I forgot to bring it back while testing. 😅

talldan commented on packages/block-editor/src/components/use-on-block-drop/index.js:
```diff
@@ -166,88 +109,134 @@ export function onFilesDrop(
+export default function useOnBlockDrop(
+	targetRootClientId,
+	targetBlockIndex,
+	replace
```
This might be good as an options object.

talldan commented on packages/block-library/src/paragraph/drop-zone.js:
```diff
@@ -0,0 +1,90 @@
+		},
+		[ clientId ]
+	);
+	const onBlockDrop = useOnBlockDrop( rootClientId, blockIndex, true );
```
I guess a similar question here about whether this should handle dragged html, files and blocks, or only files.

Possibly the component could only import `useOnFileDrop`. It might be interesting to divide the three up into hooks with compatible interfaces that can be composed or used individually.

talldan commented on packages/block-editor/src/components/use-on-block-drop/index.js:
```diff
@@ -166,88 +109,134 @@ export function onFilesDrop(
-	);
-	const { canInsertBlockType, getBlockIndex, getClientIdsOfDescendants } =
-		useSelect( blockEditorStore );
+export default function useOnBlockDrop(
```
This hook has been pretty sensitive performance wise in the past, so that's one thing to check in regards to this refactor.

I remember in the past there were issues about it not returning stable function references, and it'd trigger `useDropZone` at a high frequency. I think there was a fix in `useDropZone` though, so it may not be of concern.

kevin940726 replied:
This could use some `useEvent` for some quick wins but sadly it's not available yet 😅 .

talldan commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -55,8 +61,25 @@ function ParagraphBlock( {
-		ref: useOnEnter( { clientId, content } ),
+		ref: useMergeRefs( [
+			useOnEnter( { clientId, content } ),
+			refCallback,
```
Is the `refCallback` needed? I also wonder if `setParagraphElement` could be passed directly here as the functions have the same signature.

I guess in some ways `refCallback` might be clearer.

kevin940726 replied:
React has no guarantee that the signature of `setState` will always match the signature of the ref callback. Though I also doubt that it would ever change 😅. I'm fine either way.

talldan commented on packages/block-library/src/paragraph/editor.scss:
```diff
@@ -17,3 +17,23 @@
 }
+
+.components-popover.wp-block-paragraph__drop-zone {
+	.components-popover__content {
+		border: none;
+		outline: none;
+		box-shadow: none;
+	}
```
This is also what `BlockPopover` does, so it's probably worth making this a proper style option of the `Popover` component in a separate PR.

Popover does already have an `isAlternate` class, but I know `Button` was refactored to a `variation` prop, so it could be worth doing something like that.

I guess the other option is that there's a lower level `Floating` component that has no visible UI, and then `Popover` is the component that's presents a UI and uses `Floating`.

Marco (@ciampo) will probably have some thoughts on this.

ciampo replied:
Thank you for the ping, Dan!

Our recommendation is to avoid:

- using internal components classnames (like `.components-popover__content`) as they are not considered public APIs, and instead use classnames defined where `Popover` is consumed.
- writing style overrides to components in general, as they make it much harder for us to introduce changes to the component without causing breakages

In this case, I'm not sure if there's a way to remove border/outline/box-shadow without overrides.

Ad @talldan suggested, we could look into adding a `variant` prop to the `Popover` component (which also replaces the `isAlternate` prop), or consider alternatives (like exposing a lower level unstyled floating element).

(cc @mirka in case she has anything to add to this)

mirka replied:
> we could look into adding a `variant` props to the `Popover` component (which also replaces the `isAlternate` prop), or consider alternatives (like exposing a lower level unstyled floating element).

Agreed 👍

kevin940726 replied:
We have this pattern (overriding `.components-popover__content`) all over the codebase already 😅. Let's keep this as-is and work on a different solution as a follow-up to the Popover component improvement. WDYT?

ciampo replied:
> We have this pattern (overriding .components-popover__content) all over the codebase already 

I know, unfortunately it has been a pattern for a long time and in the past months we've been  trying to educate folks to avoid it as much as possible 😓 

Let's stick with your proposed plan for this PR

talldan commented on packages/block-library/src/paragraph/drop-zone.js:
```diff
@@ -0,0 +1,90 @@
+
+	return (
+		<Popover
+			anchor={ paragraphElement }
+			animate={ false }
+			placement="top-start"
+			focusOnMount={ false }
+			flip={ false }
+			resize={ false }
+			className="wp-block-paragraph__drop-zone"
+		>
+			<div
+				ref={ dropZoneRef }
+				className="wp-block-paragraph__drop-zone-backdrop"
+				style={ {
+					width: paragraphElement?.offsetWidth,
+					height: paragraphElement?.offsetHeight,
+				} }
+			>
+				<AnimatePresence>
+					{ isVisible ? (
+						<motion.div
+							key="drop-zone-foreground"
+							initial={
+								reducedMotion
+									? animateVariants.show
+									: animateVariants.hide
+							}
+							animate={ animateVariants.show }
+							exit={
+								reducedMotion
+									? animateVariants.show
+									: animateVariants.exit
+							}
+							className="wp-block-paragraph__drop-zone-foreground"
+						/>
+					) : null }
+				</AnimatePresence>
+			</div>
+		</Popover>
```
It's a shame that so much of this has to be implemented as a seperate component to the components' package `DropZone`, but I see why.

Do you think there's any opportunity to consolidate the two in the future? The thing with the components package version is that it's really legacy code as it implicitly calls the `useDropZone` hook (it's like that for purely back compat reasons).

I think ideally there would be some purely presentational version of the `DropZone` component.

kevin940726 replied:
> I think ideally there would be some purely presentational version of the DropZone component.

100% agreed.

talldan commented on packages/components/src/drop-zone/index.tsx:
```diff
@@ -157,7 +157,10 @@ export function DropZoneComponent( {
-				<motion.div variants={ foreground }>
+				<motion.div
+					variants={ foreground }
+					className="components-drop-zone__content-foreground"
```
Is this change and the associated css changes something that's still needed? I couldn't see that the code in this PR uses this component.

kevin940726 replied:
Oops, yeah, it's something I forgot to delete.

talldan reviewed (commented):
This looks good, code is clean and the added tests will be really helpful. Nice work. 🎉 

The main question I have is whether the empty paragraph should intercept dragged blocks as well as files, or if it should only handle files? At the moment it's handling both blocks and files, but I didn't see a discussion about that anywhere.

jasmussen commented:
> The main question I have is whether the empty paragraph should intercept dragged blocks as well as files, or if it should only handle files? At the moment it's handling both blocks and files, but I didn't see a discussion about that anywhere.

Good catch. I would lean towards it being okay to replace empty paragraphs with other blocks or patterns being dragged in there; presumably we can always pull that back if it turns out to be a problem in practice.

mtias commented:
Handling both should be fine, as we consider an empty paragraph a drop-zone in general.

kevin940726 committed:
Address code reviews

kevin940726 committed:
Minimize the refactor so that test coverage is still intact

kevin940726 committed:
Address code reviews

kevin940726 committed:
Fix the tests and refactor dragFiles util

kevin940726 commented on packages/compose/src/hooks/use-drop-zone/index.js:
```diff
@@ -219,20 +219,17 @@ export default function useDropZone( {
 			ownerDocument.addEventListener( 'dragenter', maybeDragStart );
 
 			return () => {
-				onDropRef.current = null;
-				onDragStartRef.current = null;
-				onDragEnterRef.current = null;
-				onDragLeaveRef.current = null;
-				onDragEndRef.current = null;
-				onDragOverRef.current = null;
```
This is related to the bug report in https://github.com/WordPress/gutenberg/pull/39038#discussion_r975643731. Happy to separate this into another PR if necessary.

ciampo replied:
Gave this a quick look, and I can see the mem leak again:

https://user-images.githubusercontent.com/1083581/191437281-e5c883d3-48a0-41d8-b9ba-d8d3c4ce443b.mp4

> Happy to separate this into another PR if necessary.

Up to you — it may be beneficial to tackle this separately in case this PR had to be merged urgently

kevin940726 replied:
Thanks for testing! Do you have the exact reproduction steps? I would like to give it a look!

This PR is not urgent so we can take it easy 😆.

ciampo replied:
> Do you have the exact reproduction steps?

From the video just posted above, and my original report (https://github.com/WordPress/gutenberg/pull/38999#pullrequestreview-891717184), I think we could start with:

1. Run the editor on your local machine (DEV MODE), to make sure that React fires dev-only warnings
2. Create a new post
3. Add some text to the first paragraph
4. Drag an image and drop it in the editor, under the first paragraph
5. Select the paragraph block, click and drag its drag handle and drag the paragraph under the image block
6. React would have printed the warning to console about a state update within `DropZoneComponent`

Let me know if you can repro consistently!

ciampo replied:
I actually think that this can be also reproduced on `trunk`, I saw a similar react warning when working on #44301

kevin940726 replied:
The problem seems to be that an update in one of the listeners in `useDropZone` triggers a  remount of the `<MediaPlaceholder>` component. It then triggers `useRefEffect` to fire the cleanup function. Due to a fundamental flaw in the `useRefEffect` implementation, it remounts the component **before** firing the remaining events. This usually isn't a problem, but we register another event in one of those event handlers. This causes the added event handlers to be _stalled_ and never cancel throughout the component's lifecycle.

However, if we change the `useRefEffect` implementation to `useEffect` somehow, React will make sure to run the remaining event listener before calling the cleanup effect. I think maybe this is why it's traditionally discouraged to register event listeners outside `useEffect`?

Obviously, we can't change how `useRefEffect` work now as that would impact a lot of the existing code, and I'm not aware of a way to fix it properly either. I might have some idea for this particular bug though but it won't be a general solution 😅 .

ciampo replied:
That does sound complicated 😅 

I agree that we should definitely not touch `useRefEffect` for now. 

Maybe we can try to work around the problem by re-architecting how ev listeners are added/removed in `MediaPlaceholder`, or whatever other approach you had in mind?

Also, this may have become big/complex enough to deserve its own separate PR?

kevin940726 replied:
I tried to fix it in https://github.com/WordPress/gutenberg/pull/42722/commits/49387fc6b16a45d80ca9e8da02a02f70e566aa34 by not registering/removing event listeners in another listeners, the timing they were called is probably not what we thought it would anyway.

In short, before that commit:
1. We register the `dragenter` event (on `document`) when the `ref` callback is attached to some element.
2. Once the `dragenter` event is fired, we immediately remove the event itself.
3. On `drop` or `dragend` event, we re-attach the `dragenter` event.

In https://github.com/WordPress/gutenberg/pull/42722/commits/49387fc6b16a45d80ca9e8da02a02f70e566aa34, we simply remove the last two steps and keep the listeners untouched. It would only get removed once the element is detached from DOM.

It works great in my testing, please LMK if it works for you 🙏! If you (or anyone else) think this is too complicated and deserve a separate PR, I'll be happy to do so too :).

kevin940726 replied:
Wait, I was wrong, this can still happen in `useEffect` in some cases 🤔 . Either way the commit fixes the issue though.

kevin940726 replied:
I guess the general rule of thumb is to not add event listeners inside another event listener. That could cause the listener to be stalled and never get cleaned up when the component unmounts.

ciampo replied:
> It works great in my testing, please LMK if it works for you 🙏! If you (or anyone else) think this is too complicated and deserve a separate PR, I'll be happy to do so too :).

Tested on `gutenberg.run` (as [I'm currently unable to do so locally](https://github.com/WordPress/gutenberg/issues/44416)), it seems to work well!

Definitely much cleaner! — if I had to write this piece of software, I would probably just use the `isDragging` variable in order to know what to do inside each ev listener, instead of adding/removing listeners inside listeners.

Taking a look at the code, I also noticed that [a couple of `ownerDocument.addEventListener` calls are also happening inside another ev listener](https://github.com/WordPress/gutenberg/blob/49387fc6b16a45d80ca9e8da02a02f70e566aa34/packages/compose/src/hooks/use-drop-zone/index.js#L114-L115) — should we refactor those too?

kevin940726 replied:
> should we refactor those too?

Yeah, probably! I'd do so in another PR and probably refactor `use-drop-zone` completely to hide this complexity. An idea that came to mind is to encapsulate most of the event handling in a class and just initiate that class in a `useEffect`. That way we can ensure the listeners are not out of our control when the component unmounts.

ciampo replied:
Sounds good, curious to see how this approach feels like. Please keep me in the loop when you work on that!

kevin940726 committed:
Add more e2e tests

kevin940726 commented on packages/block-editor/src/components/use-on-block-drop/index.js:
```diff
@@ -176,63 +166,145 @@ export function onFilesDrop(
+					0
+				);
+			} else {
+				moveBlocksToPosition(
```
Side note and off-topic: Ideally, we should split this into two functions of `removeBlocks` then `insertBlocks` just like the above in the replace action. So that we can allow dropping blocks from a different browser window.

talldan replied:
That's a nice idea.

kevin940726 commented on packages/block-library/src/paragraph/drop-zone.js:
```diff
@@ -0,0 +1,105 @@
+	} );
+	const [ isDragging, setIsDragging ] = useState( false );
+	const [ isVisible, setIsVisible ] = useState( false );
+	const popoverRef = useDropZone( {
+		onDragStart: () => {
+			setIsDragging( true );
+		},
+		onDragEnd: () => {
+			setIsDragging( false );
+		},
+	} );
```
We can't bind this in the `dropZoneRef` below because it hasn't been mounted yet, it's only mounted when `isDragging` is true. Ideally, we should have a unified place to listen to these events from top-down. So that unmounted children components won't affect these events (and we only have to listen to these events once/per drop zone).

kevin940726 commented on packages/block-library/src/paragraph/drop-zone.js:
```diff
@@ -0,0 +1,105 @@
+					className="wp-block-paragraph__drop-zone-backdrop"
+					ref={ dropZoneRef }
+					style={ {
+						width: paragraphElement?.offsetWidth,
+						height: paragraphElement?.offsetHeight,
```
We assume that the paragraph block won't resize itself, which is often true. Ideally though, we should leverage `useResizeObserver` once it's [ready](https://github.com/WordPress/gutenberg/pull/40509#issuecomment-1197925898) for such usage.

kevin940726 committed:
Fix dropzone memory leak

talldan commented on packages/block-editor/src/components/use-block-drop-zone/index.js:
```diff
@@ -130,8 +144,16 @@ export default function useBlockDropZone( {
 
-			if ( targetIndex !== null ) {
-				showInsertionPoint( targetRootClientId, targetIndex );
+			if ( targetIndex !== undefined && targetIndex !== null ) {
```
It's a peculiarity that `setTargetBlockIndex` falls back to `0` when `targetIndex` is `undefined`, but `showInsertionPoint` was being passed an `undefined` value.

I think `setTargetBlockIndex` is correct, as this handles dropping into an empty block list.

An insertion point (currently) can't be shown in an empty block list, so I think this change makes sense and there's no point in calling `showInsertionPoint`.

TLDR - nothing to do, this is a good improvement

kevin940726 replied:
I don't think that `targetIndex` will ever be `null` here, not sure why it was kept here. I might want to remove that check too unless there's something I'm missing?

talldan commented on packages/block-editor/src/components/use-block-drop-zone/index.js:
```diff
@@ -130,8 +144,16 @@ export default function useBlockDropZone( {
+				const nextBlock = blockElements[ targetIndex ];
+				const previousBlock = blockElements[ targetIndex - 1 ];
+
+				if (
+					( ! nextBlock || ! isEmptyParagraph( nextBlock ) ) &&
+					( ! previousBlock || ! isEmptyParagraph( previousBlock ) )
+				) {
```
I found this `if` statement pretty hard to understand. Adding a comment might help, or alternatively maybe flipping it to an early return.

In testing I was struggling to see the difference between this and an early return like this:
```js
				if (
					isEmptyParagraph( nextBlock ) ||
					isEmptyParagraph( previousBlock )
				) {
					return;
				}
```

kevin940726 replied:
Yeah, I'm not particularly happy about this either. I'll see what I can do to simplify this.

talldan commented on packages/block-editor/src/components/use-on-block-drop/index.js:
```diff
@@ -176,63 +166,145 @@ export function onFilesDrop(
+					targetBlockClientIds[ targetBlockIndex ];
+
+				// Remove the source blocks.
+				removeBlocks( sourceClientIds, false );
+				// Replace the target block with the source blocks.
+				replaceBlocks(
+					targetBlockClientId,
+					sourceBlocks,
+					undefined,
+					0
+				);
```
I wonder if these actions should be batched (`registry.batch`). (probably should, it's worth checking if undo/redo works as expected with separate action calls)

kevin940726 replied:
Oh, I totally forgot about `registry.batch`. Yes, I think these should be batched 👍.

talldan reviewed (approved):
It works really well, nice job.

Most of the comments are minor, so I have no issue approving and you can decide what to do about them.

It is a lot of added complexity to `useOnBlockDrop`, so I think in addition to the other follow-ups suggested, it's probably worth thinking about how that hook can be made easier to follow. Drag and drop is notoriously brittle, so I am a bit nervous about the scope of the changes, but I didn't spot any bugs in testing.

kevin940726 committed:
Try to fix stalled listeners issue in useDropZone

kevin940726 committed:
Address code reviews

kevin940726 committed:
Remove null check for targetIndex

youknowriad commented:
Hello folks, I didn't dive into the PR but this has a very bad impact on several performance metrics. Check https://codehealth.vercel.app

paaljoachim commented:
Thank you for working on this Kai! This will be very helpful!

Testing using Twenty Twenty Two.
WP 6.0.2

This is very helpful! Dragging and dropping an image into where the Paragraph block is.

![Drag Drop new](https://user-images.githubusercontent.com/5323259/192250990-21c8d8e5-5ae8-461e-b994-d3d0f7d6c991.jpg)


I happen to do a very quick CTRL+Z as in undoing the image upload.

It went into a loop.
![Screenshot 2022-09-26 at 12 06 17](https://user-images.githubusercontent.com/5323259/192250208-b23124ee-30fb-4780-a8f6-1a7e54e20ce7.jpg)

I waited a few minutes before refreshing/reloading the page. Then went into the revision screen to bring back the image I just CTRL+Z ed away.

It now looks like this:
![Screenshot 2022-09-26 at 12 09 01](https://user-images.githubusercontent.com/5323259/192250702-43fe5f85-01de-4ec7-9f9c-51ebbed58d89.jpg)
I will wait a few more minutes to see if the gif spinner ends by itself or if I need to again refresh the page.

A few minutes have passed. I will just save the page. Then delete the partly uploaded image.

kevin940726 commented:
@youknowriad Seems like the performance regression is from this line:

https://github.com/WordPress/gutenberg/blob/9580611aaa63b80a3ac490c11dfa83db52b923e5/packages/e2e-tests/specs/performance/post-editor.test.js#L160-L162

We're creating 1000 empty paragraph blocks to simulate some dummy blocks. However, since this PR, we're now also rendering 1000 dropzones for those 1000 empty paragraph blocks:

https://github.com/WordPress/gutenberg/blob/9580611aaa63b80a3ac490c11dfa83db52b923e5/packages/block-library/src/paragraph/edit.js#L118-L123

Here I made a choice to defer rendering the dropzone only when the paragraph block is empty, which I assume in practice should be a relatively rare case that someone would create dozens of empty paragraph blocks in the content. This assumption however isn't the case in the performance test.

FWIW, I'm not sure why we're not resetting the page between each test in the performance test suite? This causes some former metrics to alter the results of the latter ones. In this case, all metrics below `Selecting blocks` is skewed.

An obvious solution is to avoid rendering the dropzone completely when the user is not dragging. This is however not possible without some refactoring and/or requiring a `<DropZoneProvider>` to only listen to events once on the topmost level.

Another solution is to change the performance test to not render empty paragraph blocks 1000 times? Maybe simply adding some dummy text to the `content` attribute would fix this issue?

kevin940726 commented:
@paaljoachim That issue seems to be an existing one before this PR. If you drop an image between blocks using the "in-between inserter" and click "Undo", it will also enter an infinite-loading state. I _guess_ behind the scenes it's the same issue as in https://github.com/WordPress/gutenberg/pull/42001, a naughty `useEffect` is not being used correctly in the image block 😅.

youknowriad commented:
@kevin940726 That's interesting, so just the fact that we render drop zones is causing all these issues? Seems like we might have a serious problem in that component directly because even if we render 1000 drop zones, only a small number of these components is actually "synchronous" (AsyncModeProvider). 

I agree that the test is not representative but I think the problem is still an important one to solve, because a lot of blocks render drop zones (media, all container blocks), so while empty paragraphs is not common, having a lot of drop zones is.

kevin940726 commented:
@youknowriad I don't think that's a new problem introduced in this PR though, if rendering dropzones is costly then it should be like that for a while. However, I _guess_ the main bottleneck here is that we're using `<Popover>` for each dropzone to correctly position itself on top of the paragraph block (to avoid changing the DOM structure). That's 1000 popovers, which is likely what causes the page to be unresponsive. AFAIK, that's not an issue for other dropzones (`media-placeholder` for instance) because they just render a div positioned relatively.

I do have some plans to refactor drop zones to share the same state and events globally though. Maybe that would help improve the performance a bit in the long term.

youknowriad commented:
>  do have some plans to refactor drop zones to share the same state and events globally though

For what it's worth that's how drop zones were implemented originally, and it was refactored (simplified I guess) to what it is today either by @talldan or @ellatrix I don't remember exactly

ellatrix commented:
> However, I _guess_ the main bottleneck here is that we're using `<Popover>` for each dropzone to correctly position itself on top of the paragraph block (to avoid changing the DOM structure). That's 1000 popovers, which is likely what causes the page to be unresponsive.

Why are we rendering Popovers for empty paragraphs? A Popover should only be rendered if you're dragging over the block?

ellatrix commented on packages/block-library/src/paragraph/drop-zone.js:
```diff
@@ -0,0 +1,105 @@
+			className="wp-block-paragraph__drop-zone"
+			ref={ popoverRef }
+		>
+			{ isDragging ? (
```
Why render the parent element (Popover) without any content when `isDragging` is `false`?

talldan replied:
I think it's the element the dropzone references, and that's used to set `isDragging`. So it's a bit of a chicken/egg situation.

Possibly the dropzone could reference the `<p>` element.

kevin940726 replied:
Yes, `useDropZone` returns a ref callback so we need to bind it somewhere.

> Possibly the dropzone could reference the `<p>` element.

But then we'll be risking binding all those drag-and-drop listeners to the `<p>` element, regardless if it needs to render the dropzone or not, so it's more or less the same.

However, now that I think about it, maybe we only have to bind the returned ref callback to the `<p>` element when there's no `content`. I think that could work 🤔 .

A more long-term solution might be to bind the events that set `isDragging` to a global element, e.g. the root block list or the document itself. So that we don't have to think about these details when implementing DnD.

youknowriad commented:
I tried doing some performance profiling, here's the result I have for the commit before this PR

<img width="1440" alt="Screen Shot 2022-09-28 at 11 37 59 AM" src="https://user-images.githubusercontent.com/272444/192732419-cca5e1e7-ea92-4c61-9330-f1c040c9cfa9.png">

And here's the same actions (block selection) performed in the commit of the PR

<img width="1440" alt="Screen Shot 2022-09-28 at 11 40 33 AM" src="https://user-images.githubusercontent.com/272444/192732550-eaef7337-ec15-4f4a-9696-a1245c2b6d5d.png">

You can easily see that there's something constantly running in the browser, there's no idle time at all. This slows down everything. I think it doesn't matter if the issue is a pre-existing component, it's important enough to address it.

If it's hard to track, maybe we can we maybe consider reverting temporarily?

ellatrix commented on packages/block-editor/src/components/index.js:
```diff
@@ -155,6 +155,7 @@ export {
 export { default as __experimentalBlockPatternsList } from './block-patterns-list';
 export { default as __experimentalPublishDateTimePicker } from './publish-date-time-picker';
 export { default as __experimentalInspectorPopoverHeader } from './inspector-popover-header';
+export { default as __experimentalUseOnBlockDrop } from './use-on-block-drop';
```
Should we mark this unstable/internal? I don't think this is an API that we want to expose in this shape.

kevin940726 replied:
Sure. I don't mind changing it to `__internal` or any other naming. I only use `__experimental` here because that's a more established pattern.

talldan commented:
> For what it's worth that's how drop zones were implemented originally, and it was refactored (simplified I guess) to what it is today either by @talldan or @ellatrix I don't remember exactly

Yep, there was a `DropZoneProvider` that was removed from the editors in a PR by @ellatrix.

Worth noting though that it sounds like there's a desire to have some kind of top-level drop zone - https://github.com/WordPress/gutenberg/pull/43493#issuecomment-1250967222. Maybe instead of a dedicated Provider, there could be a way to check for drag over or drag start of parent drop zones. That would allow conditional rendering of the Popover. Worth noting that the block list is already a drop zone, so this is something that is already possible.

> If it's hard to track, maybe we can we maybe consider reverting temporarily?

Feels a bit early to do that. It'd be good to explore whether there are any quick fixes first. I also think it's worth looking at that performance test again, and considering whether it should be testing a post full of content, as that's the more common use case.

youknowriad commented:
>  It'd be good to explore whether there are any quick fixes first. I also think it's worth looking at that performance test again, and considering whether it should be testing a post full of content, as that's the more common use case.

Indeed, we can do the test with "large-post.html" if needed but I also noticed that even with full paragraphs, it seems there's some work that is triggered constantly on idle time of the browser less than with empty paragraphs but existent as opposed to before.

kevin940726 commented:
> I tried doing some performance profiling, here's the result I have for the commit before this PR
> 
> <img alt="Screen Shot 2022-09-28 at 11 37 59 AM" width="1440" src="https://user-images.githubusercontent.com/272444/192732419-cca5e1e7-ea92-4c61-9330-f1c040c9cfa9.png">
> 
> And here's the same actions (block selection) performed in the commit of the PR
> 
> <img alt="Screen Shot 2022-09-28 at 11 40 33 AM" width="1440" src="https://user-images.githubusercontent.com/272444/192732550-eaef7337-ec15-4f4a-9696-a1245c2b6d5d.png">

I don't think the scales of these two screenshots are the same? The duration of "before" image is ~24s while the duration of the "after" image is only ~6s. The screenshots difference might not be super representative here. Could you share the profiling files instead? Or maybe even Playwright trace file if it helps?

From my testing, I can't seem to notice any big performance regression after applying this PR, assuming there are a limited amount of empty paragraphs on the post.

youknowriad commented:
@kevin940726 These are the exact same tests, 1000 empty paragraphs like the actual performance test, the problem with the one with this branch is that it contains so much data that the browser "cut it" and didn't display the whole graph for me. And the lag is very noticeable in my machine. @mcsf can also confirm as we did this test together this morning.

ellatrix commented:
It doesn't seem necessary to me to render a Popover if we're not showing anything visually to the user. We should use the block element ref to attach a drag listener to set an `isDraggingOver` flag, and only maybe then render a Popover to indicate that it's possible to drop.

There's no need for a drag/drop provider? I don't follow the reasoning there.

Regardless of the performance impact, I think adding all this logic to the paragraph block isn't the right approach. It's making the paragraph even more complex. And what if we want to extend this to other empty/unmodified blocks? What if the default block is changed to another block? 

In my opinion, this functionality should be added either internally in `useBlockProps` or the `BlockListBlock` component, or through the `editor.BlockListBlock` filter, using `isUnmodifiedDefaultBlock` as a condition for attaching drag listeners.

kevin940726 commented:
> These are the exact same tests, 1000 empty paragraphs like the actual performance test

I agree that there's room for improvement, and I already have some plans and proposals for refactoring the drag-and-drop feature from the ground up. But I don't think an edge case like 1000 empty paragraphs is major enough for us to revert this PR, maybe I'm missing something though. It'd be nice if there's any quick win that we can do to improve it at the time.

> We should use the block element ref to attach a drag listener to set an `isDraggingOver` flag, and only maybe then render a Popover to indicate that it's possible to drop.

The problem is that the drop zone has to be rendered on top of the paragraph block. So that once we hover on the paragraph and trigger the `dragover` event, we'll have to render the Popover on top which will then trigger `dragleave` of the same component. It's still possible though using some OR logic and I think it could be a good follow-up.

> There's no need for a drag/drop provider? I don't follow the reasoning there.

Not necessarily a provider, but instead binding some common events at the top-most element only once to solve some awkward issues. I think this has been [repeated](https://github.com/WordPress/gutenberg/pull/33435#issuecomment-881728499) in the past in some issues already. I think it's worth revisiting the use cases and experimenting with them.

> And what if we want to extend this to other empty/unmodified blocks? What if the default block is changed to another block?

I can't think of any use case for that for now, but maybe I'm missing the bigger picture. I think a big change like that deserves a follow-up. I tried to build the `drop-zone` component in a modular way so it should not be too hard to remove it whenever needed.

> using `isUnmodifiedDefaultBlock` as a condition for attaching drag listeners

Ah cool! I didn't even know `isUnmodifiedDefaultBlock` is a thing before 😅. I agree that would be interesting to try it out in a follow-up.

youknowriad commented:
The performance issue here is actually related to this https://github.com/WordPress/gutenberg/pull/43617/files#r985479949 (Turning this off "solves" the performance issue)

talldan commented:
@youknowriad I did try setting that to `false` in a PR here - [Test popover animation frame performance](https://github.com/WordPress/gutenberg/pull/44604#top).

The performance test isn't showing any difference - https://github.com/WordPress/gutenberg/actions/runs/3156955991/jobs/5137248278#step:5:101. Maybe I made a mistake.

kevin940726 commented:
A heads-up that I'm currently trying to improve this in a follow-up with a much bigger scope. I'm trying to integrate it inside the `<InsertionPoint>` component since they are similar in some way and we can reuse most of the logic and positioning components there.

I'll try to share more results soon.

ellatrix commented:
@kevin940726 Ping me for a review

youknowriad commented:
> The performance test isn't showing any difference - https://github.com/WordPress/gutenberg/actions/runs/3156955991/jobs/5137248278#step:5:101. Maybe I made a mistake.

@talldan I didn't run the test so could be that I'm wrong in terms of "numbers" but when profiling, the browser is back to a decent state (idle time, no constant work being done). Your PR is going to be a great starting point to debug this properly :) I'll tell you if I find anything that stand out.

talldan commented:
> @talldan I didn't run the test so could be that I'm wrong in terms of "numbers" but when profiling, the browser is back to a decent state (idle time, no constant work being done). Your PR is going to be a great starting point to debug this properly :) I'll tell you if I find anything that stand out.

Yeah, that's exactly what I saw too, I was very surprised when there were no differences in the PR 😄

youknowriad commented:
Tried comparing the performance graphs, it seems to be that there's no "obvious" thing, it's just that we're rendering a lot more sync things. In other words, rendering the "Popover" component is intrinsically too slow. Maybe because it uses DOM functions like retrieving DOM rect... maybe there's some caching that we can do there 🤔

youknowriad commented:
Ok, so I went a rabbit whole here and I found something **very interesting**. So it turns out we have a potentially an important performance issue with how bubblesVirtually slot/fill is implemented. Basically, every time a fill gets rendered (which happens for instance when you select a block (fill for the toolbar)), all components calling `useSlot` (or using Popover as that one calls useSlot) get re-rendered synchronously (regardless of Async mode of the data provider because it's a context update, not a data module update). 

This is shown clearly in the impact this PR had but potentially, it could be already an issue in other situations as well. I'm going to see if I can come up with a fix and see what impact it has on the numbers.
